File: app/src/main/java/com/example/loginpageproject/ForgotPasswordActivity.java
package com.example.loginpageproject;

import android.content.Intent;
import android.os.Bundle;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.material.textfield.TextInputEditText;

public class ForgotPasswordActivity extends BaseActivity {

    private TextInputEditText etEmail;
    private Button btnSendCode;
    private TextView tvBackToLogin;
    private DatabaseHelper dbHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_forgot_password);

        dbHelper = new DatabaseHelper(this);
        etEmail = findViewById(R.id.etEmail);
        btnSendCode = findViewById(R.id.btnSendCode);
        tvBackToLogin = findViewById(R.id.tvBackToLogin);

        btnSendCode.setOnClickListener(v -> {
            String email = etEmail.getText().toString().trim();
            if (email.isEmpty()) {
                Toast.makeText(this, "Provide a frequency, entity.", Toast.LENGTH_SHORT).show();
                return;
            }

            if (dbHelper.checkEmailExists(email)) {
                // In a real app, send OTP. Here we skip to Reset
                Intent intent = new Intent(this, ResetPasswordActivity.class);
                intent.putExtra("email", email);
                intent.putExtra("fromForgot", true);
                startActivity(intent);
                finish();
            } else {
                Toast.makeText(this, "Frequency not found in the network.", Toast.LENGTH_SHORT).show();
            }
        });

        tvBackToLogin.setOnClickListener(v -> finish());
    }
}

File: app/src/main/kotlin/com/example/loginpageproject/ForgotPasswordActivity.kt
package com.example.loginpageproject

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.lifecycle.lifecycleScope
import com.example.loginpageproject.auth.AuthValidator
import kotlinx.coroutines.launch

class ForgotPasswordActivity : BaseActivity() {
    private lateinit var email: EditText
    private lateinit var sendButton: Button

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_forgot_password)
        email = findViewById(R.id.etEmail)
        sendButton = findViewById(R.id.btnSendCode)
        sendButton.setOnClickListener { sendCode() }
        findViewById<TextView>(R.id.tvBackToLogin).setOnClickListener { finish() }
    }

    private fun sendCode() {
        val enteredEmail = email.text.toString().trim()
        if (!AuthValidator.isValidEmail(enteredEmail)) {
            email.error = "Enter a valid registered email address"
            return
        }
        setLoading(true)
        lifecycleScope.launch {
            // Explicit existence check per spec: the account must be validated before
            // sending an OTP. Note this intentionally reveals whether an email is
            // registered (an account-enumeration trade-off accepted in favor of literal
            // spec compliance over the more conservative generic-message pattern).
            val exists = runCatching { authRepository.emailExists(enteredEmail) }.getOrElse {
                Toast.makeText(this@ForgotPasswordActivity, it.message ?: "Unable to validate email.", Toast.LENGTH_LONG).show()
                setLoading(false)
                return@launch
            }
            if (!exists) {
                email.error = "No account is registered with this email address."
                setLoading(false)
                return@launch
            }
            runCatching { authRepository.sendRecoveryOtp(enteredEmail) }
                .onSuccess {
                    Toast.makeText(this@ForgotPasswordActivity, "A reset code was sent to your email.", Toast.LENGTH_LONG).show()
                    startActivity(Intent(this@ForgotPasswordActivity, OTPActivity::class.java).apply {
                        putExtra(OTPActivity.EXTRA_EMAIL, enteredEmail)
                        putExtra(OTPActivity.EXTRA_MODE, OTPActivity.MODE_RECOVERY)
                    })
                }
                .onFailure { Toast.makeText(this@ForgotPasswordActivity, it.message ?: "Unable to send reset code.", Toast.LENGTH_LONG).show() }
            setLoading(false)
        }
    }

    private fun setLoading(loading: Boolean) {
        sendButton.isEnabled = !loading
        findViewById<View>(R.id.forgotProgress).visibility = if (loading) View.VISIBLE else View.GONE
    }
}
